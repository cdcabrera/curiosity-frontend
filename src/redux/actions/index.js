import { platformActions } from './platformActions';

const actions = {
  platform: platformActions
};

const reduxActions = { ...actions };

export { reduxActions as default, reduxActions, platformActions };
