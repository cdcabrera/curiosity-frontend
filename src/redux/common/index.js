import { reduxHelpers } from './reduxHelpers';

/**
 * @memberof Redux State
 * @module Helpers
 * @property {module} ReduxHelpers
 */

export { reduxHelpers as default, reduxHelpers };
