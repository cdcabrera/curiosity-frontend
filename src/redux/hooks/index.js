import { reactReduxHooks } from './useReactRedux';

/**
 * @memberof Redux State
 * @module Hooks
 * @property {module} UseReactRedux
 */

/**
 * Store hooks
 *
 * @type {{reactReduxHooks: {shallowEqual: Function, useDispatch: Function, useSelector: Function, useSelectors: Function,
 *     useSelectorsResponse: Function, useSelectorsAllSettledResponse: Function, useSelectorsAnyResponse: Function,
 *     useSelectorsRaceResponse: Function}}}
 */
const storeHooks = {
  reactRedux: reactReduxHooks
};

export { storeHooks as default, storeHooks, reactReduxHooks };
