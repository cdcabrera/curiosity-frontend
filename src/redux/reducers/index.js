import { combineReducers } from 'redux';
import { notifications } from '@redhat-cloud-services/frontend-components-notifications';
import appReducer from './appReducer';
import dynamicReducer from './dynamicReducer';

const reducers = {
  notifications,
  app: appReducer,
  dynamic: dynamicReducer
};

const reduxReducers = combineReducers(reducers);

export { reduxReducers as default, reduxReducers, appReducer, dynamicReducer };
