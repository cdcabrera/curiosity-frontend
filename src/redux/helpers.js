import _camelCase from 'lodash/camelCase';
import _snakeCase from 'lodash/snakeCase';
import { helpers } from '../common';

/**
 * @memberof Helpers
 * @module ReduxHelpers
 */

/**
 * Set an API query based on specific API "acceptable values" schema.
 *
 * @param {object} values
 * @param {object} schema
 * @param {*} [initialValue]
 * @returns {object}
 */
const setApiQuery = (values, schema, initialValue) => {
  const generated = {};
  const schemaArr = (schema && Object.values(schema)) || [];

  schemaArr.forEach(value => {
    if (initialValue === undefined) {
      if (value in values) {
        generated[value] = values?.[value];
      }
    } else {
      generated[value] = values?.[value] || initialValue;
    }
  });

  return generated;
};

// ToDo: research applying a maintained schema map/normalizer such as, normalizr
/**
 * Apply a set of schemas using either an array of objects in the
 * form of [{ madeUpKey: 'some_api_key' }], or an array of arrays
 * in the form of [['some_api_key','another_api_key']]
 *
 * @param {Array} schemas
 * @param {*} [initialValue]
 * @returns {Array}
 */
const setResponseSchemas = (schemas = [], initialValue) =>
  schemas.map(schema => {
    const generated = {};
    const arr = (Array.isArray(schema) && schema) || Object.values(schema);

    arr.forEach(value => {
      generated[value] = initialValue;
    });

    return generated;
  });

/**
 * Normalize an API response.
 *
 * @param {*} responses
 * @param {object} responses.response
 * @param {object} responses.response.schema
 * @param {Array|object} responses.response.data
 * @param {string} responses.response.keyCase
 * @param {Function} responses.response.customResponseEntry
 * @param {Function} responses.response.customResponseValue
 * @param {string} responses.response.keyPrefix
 * @returns {Array}
 */
const setNormalizedResponse = (...responses) => {
  const parsedResponses = [];

  responses.forEach(
    ({ schema = {}, data, customResponseEntry, customResponseValue, keyPrefix: prefix, keyCase = 'camel' }) => {
      const isArray = Array.isArray(data);
      const updatedData = (isArray && data) || (data && [data]) || [];
      const [generatedSchema = {}] = setResponseSchemas([schema]);
      const parsedResponse = [];

      updatedData.forEach((value, index) => {
        const generateReflectedData = ({
          dataObj,
          keyPrefix = '',
          keyCaseType,
          customEntry,
          customValue = null,
          update = helpers.noop
        }) => {
          let updatedDataObj = {};

          Object.entries(dataObj).forEach(([dataObjKey, dataObjValue]) => {
            let casedDataObjKey;

            switch (keyCaseType) {
              case 'camel':
                casedDataObjKey = _camelCase(`${keyPrefix} ${dataObjKey}`).trim();
                break;
              case 'snake':
                casedDataObjKey = _snakeCase(`${keyPrefix} ${dataObjKey}`).trim();
                break;
              case 'default':
              default:
                casedDataObjKey = `${dataObjKey}`.trim();
                break;
            }

            let val = dataObjValue;

            if (typeof val === 'number') {
              val = (Number.isInteger(val) && Number.parseInt(val, 10)) || Number.parseFloat(val) || val;
            }

            if (typeof customValue === 'function') {
              updatedDataObj[casedDataObjKey] = customValue({ data: dataObj, key: dataObjKey, value: val, index });
            } else {
              updatedDataObj[casedDataObjKey] = val;
            }
          });

          if (typeof customEntry === 'function') {
            updatedDataObj = customEntry(updatedDataObj, index);
          }

          update(updatedDataObj);
        };

        generateReflectedData({
          keyPrefix: prefix,
          dataObj: { ...generatedSchema, ...value },
          keyCaseType: keyCase,
          customEntry: customResponseEntry,
          customValue: customResponseValue,
          update: generatedData => parsedResponse.push(generatedData)
        });
      });

      parsedResponses.push(parsedResponse);
    }
  );

  return parsedResponses;
};

const reduxHelpers = {
  setApiQuery,
  setResponseSchemas,
  setNormalizedResponse
};

export { reduxHelpers as default, reduxHelpers };
