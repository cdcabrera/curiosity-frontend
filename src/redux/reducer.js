import _get from 'lodash/get';
import { appTypes } from '../types';
import { rhsmConstants } from '../../services/rhsm/rhsmConstants';
import { reduxHelpers } from '../common';

/**
 * Application related state reducer.
 *
 * @memberof Reducers
 * @module AppReducer
 */

/**
 * Initial state.
 *
 * @private
 * @type {{errors: {}}}
 */
const initialState = {
  errors: {}
};

/**
 * Apply application observer/reducer logic to state, against actions.
 *
 * @param {object} state
 * @param {object} action
 * @returns {object|{}}
 */
const appReducer = (state = initialState, action) => {
  switch (action.type) {
    case reduxHelpers.HTTP_STATUS_RANGE(appTypes.STATUS_4XX):
      const actionStatus = reduxHelpers.getStatusFromResults(action);

      if (actionStatus === 401 || actionStatus === 403) {
        const errorCodes = _get(reduxHelpers.getDataFromResults(action), [rhsmConstants.RHSM_API_RESPONSE_ERRORS], []);

        return reduxHelpers.setStateProp(
          'errors',
          {
            error: true,
            errorMessage: reduxHelpers.getMessageFromResults(action),
            data: errorCodes.map(value => value[rhsmConstants.RHSM_API_RESPONSE_ERRORS_TYPES.CODE]),
            status: reduxHelpers.getStatusFromResults(action)
          },
          {
            state,
            initialState
          }
        );
      }

      return state;

    default:
      return state;
  }
};

appReducer.initialState = initialState;

export { appReducer as default, initialState, appReducer };
