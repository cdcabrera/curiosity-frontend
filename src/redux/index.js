import { connect, useSelector as UseSelector } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { store } from './store';
import { reduxActions } from './actions';
import { reduxHelpers, apiQueries } from './common';
import { reduxReducers } from './reducers';
import { reduxSelectors } from './selectors';
import { reduxTypes } from './types';
import { helpers } from '../common';

const connectRouter = (mapStateToProps, mapDispatchToProps) => component =>
  withRouter(connect(mapStateToProps, mapDispatchToProps)(component));

const useSelector = (selector, value = null, options = {}) =>
  (helpers.TEST_MODE && value) || UseSelector(selector, options.equality) || value;

export {
  apiQueries,
  connect,
  connectRouter,
  reduxActions,
  reduxHelpers,
  reduxReducers,
  reduxSelectors,
  reduxTypes,
  store,
  useSelector
};
