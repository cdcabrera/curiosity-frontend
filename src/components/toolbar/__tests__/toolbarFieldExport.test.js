import React from 'react';
import {
  ToolbarFieldExport,
  toolbarFieldOptions,
  useExport,
  useExportStatus,
  useOnSelect
} from '../toolbarFieldExport';
import { store } from '../../../redux/store';
import { PLATFORM_API_EXPORT_STATUS_TYPES } from '../../../services/platform/platformConstants';

describe('ToolbarFieldExport Component', () => {
  let mockDispatch;

  beforeEach(() => {
    mockDispatch = jest.spyOn(store, 'dispatch').mockImplementation((type, data) => ({ type, data }));
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render a basic component', async () => {
    const props = {
      useExport: () => jest.fn()
    };
    const component = await shallowComponent(<ToolbarFieldExport {...props} />);

    expect(component).toMatchSnapshot('basic');
  });

  it('should export select options', () => {
    expect(toolbarFieldOptions).toMatchSnapshot('toolbarFieldOptions');
  });

  it('should handle updating export through redux state with component', () => {
    const props = {
      useOnSelect: () => jest.fn(),
      useExport: () => ({ checkAllExports: jest.fn() })
    };

    const component = renderComponent(<ToolbarFieldExport {...props} />);
    const input = component.find('button');
    component.fireEvent.click(input);

    const inputMenuItem = component.find('a.pf-v5-c-dropdown__menu-item');
    component.fireEvent.click(inputMenuItem);

    expect(mockDispatch.mock.calls).toMatchSnapshot('dispatch, component');
  });

  it('should handle updating export through redux state with hook', () => {
    const options = {
      useExport: () => ({ createExport: jest.fn() }),
      useProduct: () => ({ viewId: 'loremIpsum' }),
      useProductExportQuery: () => ({})
    };

    const onSelect = useOnSelect(options);

    onSelect({
      value: 'dolor sit'
    });
    expect(mockDispatch.mock.calls).toMatchSnapshot('dispatch, hook');
  });

  it('should aggregate export status, polling status with a hook', async () => {
    const { result: basic, unmount: unmountBasic } = await renderHook(() =>
      useExportStatus({
        useProduct: () => ({
          productId: 'loremIpsum'
        })
      })
    );
    await unmountBasic();
    expect(basic).toMatchSnapshot('status, basic');

    const { result: polling, unmount: unmountPolling } = await renderHook(() =>
      useExportStatus({
        useProduct: () => ({
          productId: 'loremIpsum'
        }),
        useSelector: () => ({
          isPending: true,
          pending: [
            {
              status: PLATFORM_API_EXPORT_STATUS_TYPES.PENDING,
              format: 'dolorSit'
            }
          ]
        })
      })
    );
    await unmountPolling();
    expect(polling).toMatchSnapshot('status, polling');

    const { result: completed, unmount: unmountCompleted } = await renderHook(() =>
      useExportStatus({
        useProduct: () => ({
          productId: 'loremIpsum'
        }),
        useSelector: () => ({
          isPending: false,
          pending: []
        })
      })
    );
    await unmountCompleted();
    expect(completed).toMatchSnapshot('status, completed');
  });

  it('should aggregate export service calls', async () => {
    // confirm attempt at creating an export
    const mockServiceCreateExport = jest.fn().mockImplementation(
      (...args) =>
        dispatch =>
          dispatch(...args)
    );
    const {
      result: { createExport },
      unmount: unmountCreate
    } = await renderHook(() =>
      useExport({
        createExport: mockServiceCreateExport
      })
    );
    createExport('mock-product-id', { data: { lorem: 'ipsum' } });
    await unmountCreate();
    expect(mockServiceCreateExport).toHaveBeenCalledTimes(1);
    expect(mockServiceCreateExport.mock.calls).toMatchSnapshot('createExport');

    // confirm attempt at getting an existing export status
    const mockServiceGetExistingExportsStatus = jest.fn().mockImplementation(
      (...args) =>
        dispatch =>
          dispatch(...args)
    );
    const {
      result: { checkAllExports },
      unmount: unmountStatus
    } = await renderHook(() =>
      useExport({
        getExistingExportsStatus: mockServiceGetExistingExportsStatus
      })
    );
    checkAllExports();
    await unmountStatus();
    expect(mockServiceGetExistingExportsStatus).toHaveBeenCalledTimes(1);
  });
});
