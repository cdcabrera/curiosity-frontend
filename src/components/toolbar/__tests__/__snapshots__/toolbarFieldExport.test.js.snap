// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToolbarFieldExport Component should aggregate export status, polling status with a hook: status, basic 1`] = `
{
  "isProductPending": false,
  "pendingProductFormats": [],
}
`;

exports[`ToolbarFieldExport Component should aggregate export status, polling status with a hook: status, completed 1`] = `
{
  "isProductPending": false,
  "pendingProductFormats": [],
}
`;

exports[`ToolbarFieldExport Component should aggregate export status, polling status with a hook: status, polling 1`] = `
{
  "isProductPending": true,
  "pendingProductFormats": [
    "dolorSit",
  ],
}
`;

exports[`ToolbarFieldExport Component should allow export service calls on existing exports: existingExports 1`] = `
[
  {
    "type": "return",
    "value": {
      "rejected": {
        "description": "t(curiosity-toolbar.notifications_export_error, {"context":"description"})",
        "dismissable": true,
        "title": "t(curiosity-toolbar.notifications_export_error, {"context":"title"})",
        "variant": "warning",
      },
    },
  },
  {
    "type": "return",
    "value": {
      "payload": "swatch-exports-status",
      "type": "@@INSIGHTS-CORE/NOTIFICATIONS/REMOVE_NOTIFICATION",
    },
  },
  {
    "type": "return",
    "value": {
      "payload": {
        "autoDismiss": false,
        "description": <div
          aria-live="polite"
        >
          t(curiosity-toolbar.notifications_export_completed_description_existing, {"context":"pending","count":1,"completed":0,"pending":1})
          <p
            style={
              {
                "paddingTop": "1em",
              }
            }
          >
            <Button
              autoFocus={true}
              data-test="exportButtonConfirm"
              onClick={[Function]}
              variant="primary"
            >
              t(curiosity-toolbar.button, {"context":"yes"})
            </Button>
             
            <Button
              data-test="exportButtonConfirm"
              onClick={[Function]}
              variant="plain"
            >
              t(curiosity-toolbar.button, {"context":"no"})
            </Button>
          </p>
        </div>,
        "dismissable": false,
        "id": "swatch-exports-status",
        "title": "t(curiosity-toolbar.notifications_export_completed_title, {"context":"existing","count":1})",
      },
      "type": "@@INSIGHTS-CORE/NOTIFICATIONS/ADD_NOTIFICATION",
    },
  },
]
`;

exports[`ToolbarFieldExport Component should allow export service calls: createExport 1`] = `
[
  {
    "type": "return",
    "value": {
      "id": "mock-product-id",
      "isPending": true,
      "type": "SET_PLATFORM_EXPORT_STATUS",
    },
  },
  {
    "type": "return",
    "value": "mock-product-id",
  },
]
`;

exports[`ToolbarFieldExport Component should export select options: toolbarFieldOptions 1`] = `
[
  {
    "selected": false,
    "title": "t(curiosity-toolbar.label_export, {"context":"json"})",
    "value": "json",
  },
]
`;

exports[`ToolbarFieldExport Component should handle updating export through redux action with hook: dispatch, hook 1`] = `
[
  [
    undefined,
    {
      "expires_at": "2019-07-21T00:00:00.000Z",
      "format": "dolor sit",
      "name": "swatch-undefined",
      "sources": [
        {
          "application": "subscriptions",
          "filters": {},
          "resource": "subscriptions",
        },
      ],
    },
  ],
]
`;

exports[`ToolbarFieldExport Component should render a basic component: basic 1`] = `
<Select
  aria-label="t(curiosity-toolbar.placeholder, {"context":"export"})"
  ariaLabel="Select option"
  buttonVariant="plain"
  className=""
  data-test="toolbarFieldExport"
  direction="down"
  id="generatedid-"
  isDisabled={false}
  isDropdownButton={true}
  isFlipEnabled={false}
  isInline={true}
  isToggleText={true}
  maxHeight={null}
  name={null}
  onSelect={[Function]}
  onSplitButton={[Function]}
  options={
    [
      {
        "isDisabled": false,
        "selected": false,
        "title": "t(curiosity-toolbar.label_export, {"context":"json"})",
        "value": "json",
      },
    ]
  }
  placeholder="t(curiosity-toolbar.placeholder, {"context":"export"})"
  position="left"
  selectedOptions={null}
  splitButtonVariant={null}
  title="t(curiosity-toolbar.placeholder, {"context":"export"})"
  toggleIcon={<ExportIcon />}
  variant="single"
/>
`;
